### PR TITLE
Cached Auto-Login + Profile Screen Improvements

### PR DESCRIPTION
**Pull request by thisjustin123**
**Changes:**

- Automatically logs the user in when they have already logged in before.
- Saves username and password to DataStore with the password encrypted (AES-GCM encryption with AndroidKeyStore).
- Displays cached user account info during auto-login. When login finishes, data is updated to reflect actual values. (User info is cached to DataStore upon every successful login).
- Meal Swipes text no longer always says " ... remaining this semester," and now reflects plans that handle swipes weekly, semesterly, and unlimited...ly.
- Meal Plan text now displays the user's meal plan's name instead of just "Meal Plan."
- Meal Swipes transaction history now populates with meal swipe history correctly.
- user_prefs.proto updated with a lot more properties for storage.

image
[Youtube Demo](https://www.youtube.com/watch?v=ZxnNVvvTpao)
^ Click to go to youtube demo

### CODE REVIEW

File: app/src/main/java/com/appdev/eateryblueandroid/ui/screens/ProfileScreen.kt [hunk 1]
```diff
@@ -68,20 +86,30 @@ fun ProfileScreen(
...
                         accountFilter = it.accountFilter,
                         transactionQuery = it.query,
                         updateQuery = profileViewModel::updateQuery,
                         showBottomSheet = showBottomSheet
                     )
-                is ProfileViewModel.State.LoggingIn ->
-                    Text("FAILURE STATE")
+                }
+                is ProfileViewModel.State.LoggingIn -> {
+                    //Text("FAILURE STATE")
```
Comment (kevinsun-dev) [inline app/src/main/java/com/appdev/eateryblueandroid/ui/screens/ProfileScreen.kt:99]: Consider removing this comment

File: app/src/main/java/com/appdev/eateryblueandroid/models/User.kt [hunk 2]
```diff
@@ -46,6 +48,17 @@ enum class AccountType {
...
 }
 
+
+enum class SwipesType {
+    WEEKLY, SEMESTERLY, UNLIMITED,
+
+    /**
+     * Use if this meal plan does not have meal swipes or would not be concerned with how many
+     * swipes they have. Default.
+     */
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/models/User.kt:58]: Can we even out whitespace here

File: app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import java.time.LocalDateTime
+
+var loadedUsername: String? = null
+var loadedPassword: String? = null
+var hasLoaded = false
+
+fun initializeLoginData() {
+    /*fun tryLoad() {
```
Comment (kevinsun-dev) [inline app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt:21]: If not necessary, remove block comments.

File: app/src/main/java/com/appdev/eateryblueandroid/util/Constants.kt [hunk 4]
```diff
@@ -15,4 +20,36 @@ object Constants {
     const val DATA_STORE_FILE_NAME = "user_prefs.pb"
 
     var currentLocation: Location? = null
+
+    val Context.userPreferencesStore: DataStore<UserPreferences> by dataStore(
+        fileName = DATA_STORE_FILE_NAME,
+        serializer = UserPreferencesSerializer
+    )
+
+    val passwordAlias = "passwordy password"
```
Comment (kevinsun-dev) [inline app/src/main/java/com/appdev/eateryblueandroid/util/Constants.kt:29]: ... right. I'm just going to leave that judgement call to you.

File: app/src/main/proto/user_prefs.proto [hunk 5]
```diff
@@ -4,6 +4,44 @@ option java_package = "com.codelab.android.datastore";
...
+message AccountProto {
+  int32 type = 1;
+  double balance = 2;
+}
+
+message TransactionProto {
+  string id = 1;
+  double amount = 2;
+  double resultingBalance = 3;
+  // date here
```
Comment (kevinsun-dev) [inline app/src/main/proto/user_prefs.proto:33]: That's a bit of a self-describing comment.

File: app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            Log.i("IO", "Username and password loaded: $username, $password")
+            hasLoaded = true
+            this.cancel()
+        }
+        if (username.isEmpty() && password.isEmpty()) {
+            hasLoaded = true
+            this.cancel()
+        }
+    }*/
+
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt:36]: Why all this commented code here?

File: app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                this.cancel()
+            }
+        }
+    }
+    CoroutineScope(Dispatchers.IO).launch {
+        passwordFlow.collect { encryptedPass ->
+            //Use password...
+            loadedPassword = encryptedPass
+            if (encryptedPass.isNotEmpty()) {
+                //tryLoad()
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt:60]: commented code?

File: app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    if (!CachedAccountInfo.cached) return User()
+
+    return User(
+        id = loadedUsername,
+        accounts = CachedAccountInfo.accounts,
+        transactions = CachedAccountInfo.transactions,
+        swipesType = CachedAccountInfo.swipesType,
+        mealPlanName = CachedAccountInfo.mealPlanName
+    )
+}
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/util/AutoLoginUtils.kt:255]: Add newline at end of file

File: app/src/main/java/com/appdev/eateryblueandroid/util/Constants.kt [hunk 9]
```diff
@@ -15,4 +20,36 @@ object Constants {
...
+        "off-campus value" to "Off-Campus Value",
+        "bear traditional" to "Bear Traditional",
+        "unlimited" to "Unlimited",
+        "bear basic" to "Bear Basic",
+        "bear choice" to "Bear Choice",
+        "house meal plan" to "House",
+        "house affiliate" to "House Affiliate",
+        "flex" to "Flex 10/500",
+        "bucks" to "Just Bucks")
 }
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/util/Constants.kt:55]: Newline

File: app/src/main/java/com/appdev/eateryblueandroid/util/EncryptionUtils.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    val ivString = Base64.decode(data.substring(0, data.indexOf("||")), Base64.NO_WRAP)
+    //Log.i("CipherTest", "Grabbed IV: " + String(ivString, utf8))
+
+    // Get the actually encrypted part as a ByteArray.
+    val encryptedData = Base64.decode(data.substring(data.indexOf("||")+2), Base64.NO_WRAP)
+    //Log.i("CipherTest", encryptedData)
+
+    cipher.init(Cipher.DECRYPT_MODE, getSecretKey(alias), GCMParameterSpec(128, ivString))
+    return cipher.doFinal(encryptedData).toString(utf8)
+}
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/util/EncryptionUtils.kt:56]: Newline

File: app/src/main/java/com/appdev/eateryblueandroid/util/RecentSearchesUtils.kt [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    CoroutineScope(Dispatchers.IO).launch {
+        appContext!!.userPreferencesStore.updateData { currentPreferences ->
+            currentPreferences.toBuilder()
+                .clearRecentSearches()
+                .addAllRecentSearches(recentSearches)
+                .build()
+        }
+    }
+}
```
Comment (spd75) [inline app/src/main/java/com/appdev/eateryblueandroid/util/RecentSearchesUtils.kt:76]: Newline

File: app/src/main/proto/user_prefs.proto [hunk 12]
```diff
@@ -4,6 +4,44 @@ option java_package = "com.codelab.android.datastore";
...
+  string location = 6;
+}
+
+message Date {
+  int32 year = 1;
+  int32 month = 2;
+  int32 day = 3;
+  int32 hour = 4;
+  int32 minute = 5;
 }
```
Comment (spd75) [inline app/src/main/proto/user_prefs.proto:47]: Newline

**Review by spd75 — COMMENTED** [review]
Looks good just some nitpicks

File: app/src/main/java/com/appdev/eateryblueandroid/util/RecentSearchesUtils.kt [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * A list of ints corresponding to the recent ID's that the user has
+ * searched and clicked on.
+ *
+ * Use saveRecentSearch(search) to add recent searches to the list.
+ *
+ * Index 0 corresponds to the newest search, index size-1 is the oldest search.
+ */
+val recentSearches : List<Int> = mutableListOf()
```
Comment (kevinsun-dev) [inline app/src/main/java/com/appdev/eateryblueandroid/util/RecentSearchesUtils.kt:22]: You seem to cast the list a lot. Instead, do you want to just set the type of this to MutableList?

**Review by kevinsun-dev — APPROVED** [review]
Small changes requested, looks great overall.

**Pull request merged**